### PR TITLE
[OTA] Implement event support for OTA Requestor cluster

### LINKED ISSUE #13006: OTA Requestor events integration
#### Problem
There is no logging of OTA Requestor events

 #### Proposed Solution
Call `PlatformMgr().SetDelegate` and log OTA Requestor events whenever appropriate using `LogEvent`

### PR DESCRIPTION
**Pull request by carol-apple**
#### Problem
Events are not logged for OTA Requestor cluster

Fixes: https://github.com/project-chip/connectedhomeip/issues/13006

#### Change overview
- Add mechanism to log the following events OTA Requestor cluster events:
-- StateTransition
-- VersionApplied
-- DownloadError
- Log the events throughout OTA Requestor core logic

#### Testing
Manual testing for Linux provider/requestor and verify that messages like the following are seen:
```
[1642621838874] [72717:6427314] CHIP: [EVL] LogEvent event number: 0x0000000000000000 priority: 1, endpoint id:  0x0 cluster id: 0x0000_002A event id: 0x0 Sys timestamp: 0x000000003E39F24B
[1642621842882] [72717:6427314] CHIP: [EVL] LogEvent event number: 0x0000000000000003 priority: 2, endpoint id:  0x0 cluster id: 0x0000_002A event id: 0x1 Sys timestamp: 0x000000003E3A01F3
```

### CODE REVIEW

File: src/app/clusters/ota-requestor/BDXDownloader.cpp [hunk 1]
```diff
@@ -131,7 +134,7 @@ void BDXDownloader::EndDownload(CHIP_ERROR reason)
         {
             mImageProcessor->Abort();
         }
-        SetState(State::kIdle);
+        SetState(State::kIdle, OTAChangeReasonEnum::kSuccess);
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/BDXDownloader.cpp:137]: `SetState` eventually calls into OTA Requestor core to let it know the download state. When the download is ending (but not complete yet), it uses the same `kIdle` state but the core logic needs a way to differentiate when this is a successful end vs one of the above failure cases so I added in an extra param to help with that.

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 2]
```diff
@@ -129,13 +119,13 @@ void OTARequestor::OnQueryImageResponse(void * context, const QueryImageResponse
 
         if (err != CHIP_NO_ERROR)
         {
-            requestorCore->mOtaRequestorDriver->HandleError(UpdateFailureState::kQuerying, err);
+            requestorCore->RecordErrorUpdateState(UpdateFailureState::kQuerying, err);
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:122]: Added a utility `RecordErrorUpdateState` which would call `HandleError`, log state transitions as well as update server attributes. Anywhere we used to call `HandleError` will now be replaced by this function.

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 3]
```diff
@@ -183,11 +171,11 @@ void OTARequestor::OnApplyUpdateResponse(void * context, const ApplyUpdateRespon
         break;
     case OTAApplyUpdateAction::kAwaitNextAction:
         requestorCore->mOtaRequestorDriver->UpdateSuspended(System::Clock::Seconds32(response.delayedActionTime));
-        SetUpdateStateAttribute(OTAUpdateStateEnum::kDelayedOnApply);
+        requestorCore->RecordNewUpdateState(OTAUpdateStateEnum::kDelayedOnApply, OTAChangeReasonEnum::kDelayByProvider);
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:174]: Likewise, `RecordNewUpdateState` is a utility function which logs the state transition as well as update the server attribute so all previous explicit calls to `SetUpdateStateAttribute` will be replaced by this function.

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp [hunk 4]
```diff
@@ -112,7 +113,9 @@ EmberAfStatus OtaRequestorServerSetUpdateStateProgress(uint8_t value)
         app::DataModel::Nullable<uint8_t> currentValue;
         status = Attributes::UpdateStateProgress::Get(endpoint, currentValue);
         VerifyOrDie(EMBER_ZCL_STATUS_SUCCESS == status);
-        if (currentValue.IsNull() || currentValue.Value() != value)
+
+        if ((currentValue.IsNull() && !value.IsNull()) || (!currentValue.IsNull() && value.IsNull()) ||
+            (!currentValue.IsNull() && !value.IsNull() && (currentValue.Value() != value.Value())))
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/ota-requestor-server.cpp:118]: Is this just working around Nullable not having an operator==?  How about we just add an `operator==` there, which forwards to the `Optional::operator==`?

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp [hunk 5]
```diff
@@ -122,6 +125,62 @@ EmberAfStatus OtaRequestorServerSetUpdateStateProgress(uint8_t value)
...
+void OtaRequestorServerOnStateTransition(DataModel::Nullable<OTAUpdateStateEnum> previousState, OTAUpdateStateEnum newState,
+                                         OTAChangeReasonEnum reason, DataModel::Nullable<uint32_t> targetSoftwareVersion)
+{
+    if (!previousState.IsNull() && previousState.Value() == newState)
+    {
+        ChipLogError(Zcl, "Previous state and new state are the same, no event to log");
+        return;
+    }
+
+    // Find all endpoints that has OtaSoftwareUpdateRequestor implemented
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/ota-requestor-server.cpp:137]:
```suggestion
    // Find all endpoints that have OtaSoftwareUpdateRequestor implemented
```

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 6]
```diff
@@ -402,12 +385,21 @@ void OTARequestor::ApplyUpdate()
     ConnectToProvider(kApplyUpdate);
 }
 
-void OTARequestor::NotifyUpdateApplied()
+void OTARequestor::NotifyUpdateApplied(uint32_t version)
 {
+    // New version is executing so update where applicable
+    VerifyOrReturn(Basic::Attributes::SoftwareVersion::Set(kRootEndpointId, version) == EMBER_ZCL_STATUS_SUCCESS);
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:391]: @Damian-Nordic This kind of continues on our last convo on this topic. The idea is the caller (be it the application or in Linux case, the dummy ImageProcessor impl) will be the one that knows the new version that has booted and will call into OTA Requestor core to inform of this update and it'll be the core's job to update it's internal version as well as the Basic cluster attribute.

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp [hunk 7]
```diff
@@ -122,6 +125,62 @@ EmberAfStatus OtaRequestorServerSetUpdateStateProgress(uint8_t value)
...
+        if (CHIP_NO_ERROR != err)
+        {
+            ChipLogError(Zcl, "Failed to record StateTransition event: %" CHIP_ERROR_FORMAT, err.Format());
+        }
+    }
+}
+
+void OtaRequestorServerOnVersionApplied(uint32_t softwareVersion, uint16_t productId)
+{
+    // Find all endpoints that has OtaSoftwareUpdateRequestor implemented
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/ota-requestor-server.cpp:153]:
```suggestion
    // Find all endpoints that have OtaSoftwareUpdateRequestor implemented
```

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp [hunk 8]
```diff
@@ -122,6 +125,62 @@ EmberAfStatus OtaRequestorServerSetUpdateStateProgress(uint8_t value)
...
+        {
+            ChipLogError(Zcl, "Failed to record VersionApplied event: %" CHIP_ERROR_FORMAT, err.Format());
+        }
+    }
+}
+
+void OtaRequestorServerOnDownloadError(uint32_t softwareVersion, uint64_t bytesDownloaded,
+                                       DataModel::Nullable<uint8_t> progressPercent, DataModel::Nullable<int64_t> platformCode)
+{
+    // Find all endpoints that has OtaSoftwareUpdateRequestor implemented
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/ota-requestor-server.cpp:170]:
```suggestion
    // Find all endpoints that have OtaSoftwareUpdateRequestor implemented
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp (hunk 4 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/ota-requestor-server.cpp:118]: Do you want me to add `operator==` in this PR or in a separate one?

Comment (bzbarsky-apple) [reply src/app/clusters/ota-requestor/ota-requestor-server.cpp:118]: Whichever you prefer.

**Review by selissia — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**